modules/system-disaster-recovery/src/main/java/org/apache/ignite/internal/disaster/system/SystemDisasterRecoveryStorage.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.disaster.system;
+
+import org.apache.ignite.internal.cluster.management.ClusterState;
+import org.apache.ignite.internal.disaster.system.message.ResetClusterMessage;
+import org.apache.ignite.internal.disaster.system.storage.ClusterResetStorage;
+import org.apache.ignite.internal.lang.ByteArray;
+import org.apache.ignite.internal.util.ByteUtils;
+import org.apache.ignite.internal.vault.VaultEntry;
+import org.apache.ignite.internal.vault.VaultManager;
+import org.jetbrains.annotations.Nullable;
+
+class SystemDisasterRecoveryStorage extends ClusterResetStorage {
+    private static final String NODE_INITIALIZED_VAULT_KEY = "systemRecovery.nodeInitialized";
+    private static final String CLUSTER_STATE_VAULT_KEY = "systemRecovery.clusterState";
+
+    private final VaultManager vault;
+
+    SystemDisasterRecoveryStorage(VaultManager vault) {
+        super(vault);
+
+        this.vault = vault;
+    }
+
+    @Nullable ClusterState readClusterState() {
```

Comment:
We also have a local state storage in CMG manager that stores a part of the cluster state. Can we unify this somehow?

Reply:
ClusterState stored in the 'local state storage' might be destroyed (for example, when applying a ResetClusterMessage on startup), so we cannot rely on it. After such a destruction, a node cannot serve as a Conductor anymore. If we make a few failing CMG repair attempts and destroy this local state everywhere, we are in trouble.

This is why another copy of ClusterState (that never gets destroyed) is added for disaster recovery.

If your question is about unifying code, I'm not sure we should do it as these have different uses.